src/fsw/FCCode/Drivers/Piksi.cpp

```diff
@@ -26,18 +26,56 @@ sbp_msg_callbacks_node_t Piksi::_user_data_callback_node;
 #ifndef DESKTOP
 Piksi::Piksi(const std::string &name, HardwareSerial &serial_port)
     : Device(name), _serial_port(serial_port) {}
+IntervalTimer check_buffer_timer = IntervalTimer();
 #else
 Piksi::Piksi(const std::string &name) {
     _read_return = 2; // this is the no fix return condition
 }
 #endif
 
+static volatile int last_bytes = 0;
+static volatile unsigned long interrupt_count = 0;
+static volatile unsigned long start_time;
+static volatile unsigned long last_time;
+
+void Piksi::check_bytes(){
+    #ifndef DESKTOP
+    interrupt_count++;
+    int bytes = Serial4.available();
+    unsigned long time = interrupt_count;
+
+    // if bytes are entering the buffer, update the timestamp
+    if (bytes > last_bytes) {
+
+        if (time - last_time > 100){
+            start_time = time;
+        }
+
+    last_bytes = bytes;
+    last_time = time;
+   }
+
+    #endif
+}

```

Comment:
Just noting here that we have the potential to overflow an unsigned integer here. Now this was an issue with the `micros()` timer before but now we actually have the opportunity to fix it. Setting the interrupt count back to zero when we used to rewrite start time could help avoid this issue. Something like this:
```suggestion
static volatile int last_bytes = 0;
static volatile unsigned long interrupt_count = 0;
static volatile unsigned long last_interrupt_count;

void Piksi::check_bytes(){
#ifndef DESKTOP
    interrupt_count++;
    int bytes = Serial4.available();
    unsigned long count = interrupt_count;

    // if bytes are entering the buffer, update the timestamp
    if (bytes > last_bytes) {
        if (++interrupt_count - last_time > 100) {
            interrupt_count = 0;
        }

        last_bytes = bytes;
        last_interrupt_count = interrupt_count;
   }
#endif
}

```
We'd then have to update I think the calculation of `piksi.microdelta` at little bit as well. This way though we'd be completely guarded against overflows and we could even move back to `unsigned int` types to make the serializer's happy again!